### DISCUSSION TITLE
Kepler performance Analysis

### DISCUSSION THREAD

**Discussion by marceloamaral** [sustainable-computing-io/kepler#391] [Show and tell]
# Kepler Performance Analysis

## Goal
Issue #365 pointed to a possible performance regression after changing metrics from per-pod to per-container.

This discussion aims to shed light on Kepler's performance.

To understand performance, let's look at three different code versions:
- v0.3 (by pod metrics)
- before gpu update (before PR  #345)
- with gpu update (this update improved some parts of the code, that is, it started to use more pointers instead of duplicate data in arrays)

For the analysis, we are going to use the `go pprof tool` for profiling CPU and MEM (heap) analysis for 60s.

## CPU analysis
**v0.3 (per pod metrics)**
`go tool pprof cpu.prof`
<img width="652" alt="image" src="https://user-images.githubusercontent.com/5133121/201896146-65d5432c-6742-4760-8906-cf034b7b14f7.png">

**before gpu update**
`go tool pprof cpu.prof`
<img width="623" alt="image" src="https://user-images.githubusercontent.com/5133121/201896738-cc6ccaf6-4294-4de3-809b-a9e830cef589.png">

**with gpu update**
`go tool pprof cpu.prof`
<img width="656" alt="image" src="https://user-images.githubusercontent.com/5133121/201895903-835f1485-79d5-4d73-8f9a-c5446cb4cf68.png">

### CPU Analysis Comments
We can see that the number of seconds of the `runtime.cgocall` func execution has increased after the update.
In fact, we may find a lot of runtime activity, which often indicates Garbage Collector (GC) activity.
This is also the conclusion from issue #381 that the GC can be the root cause of performance degradation.

There are several reasons why the GC can become more intensive and the main one is because it must free up the heap memory of many objects.

Heap scape is a known issue in our code, we have a test that always fails and we need to improve this in the future.
However, we need to understand which functions are creating the most objects and what is happening in the code.
We'll discuss this in the next section.

Note that to mitigate the GC problem we can decrease the number of heap memory and the possibility to slow down the GC using the GOGC variable. But GOGC increases program memory usage.

## Memory analysis
**v0.3 (per pod metrics)**
`go tool pprof -alloc_objects mem.prof`
<img width="1114" alt="image" src="https://user-images.githubusercontent.com/5133121/201901591-bba2a65e-0b59-4eba-ae57-9ae0dfdf57b0.png">
<img width="1030" alt="image" src="https://user-images.githubusercontent.com/5133121/201902279-ad3a8a10-1467-4b85-9f64-0c6c0486e04e.png">

**before gpu update**
`go tool pprof -alloc_objects mem.prof`
<img width="1196" alt="image" src="https://user-images.githubusercontent.com/5133121/201901710-c76ec35c-55e4-45c6-b84f-003d30ac67f4.png">
<img width="614" alt="image" src="https://user-images.githubusercontent.com/5133121/201902603-39c53461-37ed-48d0-96f7-d2302ef65c3c.png">

**with gpu update**
`go tool pprof -alloc_objects mem.prof`
<img width="1095" alt="image" src="https://user-images.githubusercontent.com/5133121/201901478-61d143a5-fa85-4835-b2b4-81c04ad23ff4.png">
<img width="948" alt="image" src="https://user-images.githubusercontent.com/5133121/201902426-7a1fb389-dfaa-4baf-9d89-4fd57f5122cc.png">

### MEM Analysis Comments
We can see in the results that functions that are allocating more objects are: `getCPUCoreFrequency` and `kubelet ListPods`.

| Code Version | `getCPUCoreFrequency`  | `kubelet ListPods` | ListPods Ratio (version_val/v0.3_val) |
| --------------- | --------------- | --------------- |--------------- |
| v0.3 (per pod metrics) | 1082414 |    74169 | 
| before gpu update        | 1141056 | 243323| 3.2 |
| with gpu update            | 848989 | 133082 | 1.8 |

In a high-level analysis, `ListPods` appears to be the critical function that is creating more objects after the upgrade, i.e., creating up to 3.2x more objects than version v0.3.
ListPods is called when the pod information is not in the cache, so we need to call the kubelet API to get this information.
Notice that in ListPods, we iterate through the list of all pods to find the pods with the target id. Therefore, this function will become more expensive when the system has more pods running.
While we need a deeper look into this, the issue could be related to caches. So maybe we need to improve that.

Another solution we discussed earlier is the introduction of a new component/proxy that will watch for resources (pods, jobs, etc.). This proxy will act as a cache and will transmit event updates to kepler instances, where events will be filtered by node name.
The main reason to create a proxy for apiserver is to avoid overloading it with the List operation when multiple instances of kepler are restarted. With this solution, we could remove the ListPods and ensure consistency in the code as we will be able to promptly identify the deleted pods.

## Latency analysis
The logs show how long kepler took to update all metrics and the `num` is the number of pod or containers, depending on the code.
**v0.3 (per pod metrics)**
<img width="771" alt="image" src="https://user-images.githubusercontent.com/5133121/201911379-6b5189bc-9e26-48ef-b901-56243c65c43b.png">

**with gpu update**
<img width="865" alt="image" src="https://user-images.githubusercontent.com/5133121/201911313-708a57ab-cdb6-40f2-a23a-c8617302dbdf.png">

### Latency Analysis Comments
Sometimes the updated code is taking longer to update all the metrics. We need a deeper analysis to fully understand why.
But the number of containers is probably not the reason for the performance degradation as the difference in num is minimal

---
Signed-off-by: Marcelo Amaral [marcelo.amaral1@ibm.com](mailto:marcelo.amaral1@ibm.com)

**Comment by rootfs** [comment]
`ListPods` is usually linked to new Pods creation. If during the test, more pods are added (e.g. by scaling up the replica), the impact could be more visible.

**Comment by marceloamaral** [comment]
I updated method to collect CPU frequency using BPF instead of reading kernel files.
The test is compared to the most recent main, with commit `afd9d6da765c6b0c90207ed253c78ee975c71fd2`.

This update has a big performance improvement.
- Before the update, Kepler was using ~12% CPU and dropped to ~1.5%.
- The goprof CPU profile was showing ~2.23s for `runtime.cgocall`, after the update it dropped to 1.21s.
- goprof CPU profiling was showing huge heap object creation for getCPUCoreFrequency and os.newFile, update removed getCPUCoreFrequency and disabled os.newFile from kernel read CPU frequency file with available BPF.

**Comment by marceloamaral** [comment]
The next steps are to improve the `ListPods` and `prometheus MakeLabePairs`.

go tool pprof -alloc_objects  mem.prof
<img width="1149" alt="image" src="https://user-images.githubusercontent.com/5133121/204549749-c9b7afdc-0ee4-4d2e-9efe-59d2aef43a86.png">


<img width="1640" alt="image" src="https://user-images.githubusercontent.com/5133121/204549421-b3e1dfeb-e530-4b7e-aa76-c944b5b13851.png">
